core/blockchain_impl.go

```diff
@@ -781,13 +781,37 @@ func (bc *BlockChainImpl) resetWithGenesisBlock(genesis *types.Block) error {
 	return nil
 }
 
+func (bc *BlockChainImpl) repairRecreateStateTries(head **types.Block) error {
```

Comment:
This should be ok, because blocks we previously checked, so we can recreate state.